src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexTreeAnalyzer.cs

```diff
@@ -0,0 +1,169 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Threading;
+
+namespace System.Text.RegularExpressions
+{
+    /// <summary>Analyzes a <see cref="RegexTree"/> of <see cref="RegexNode"/>s to produce data on the tree structure, in particular in support of code generation.</summary>
+    internal static class RegexTreeAnalyzer
+    {
+        /// <summary>Analyzes a <see cref="RegexCode"/> to learn about the structure of the tree.</summary>
+        public static AnalysisResults Analyze(RegexCode code)
```

Comment:
Would it be possible to use the RegexTree here instead of the RegexCode? My question stems more from an offline chat we once had where you said that in the future we may want to stop invoking the RegexWriter and calculate the code unless we were on the interpreted engine, so if possible, it may be benefitial to remove other engines dependencies on the code as opposed to adding new ones. This of course could be done later as well, but I'm just curious since by looking at this file, it doesn't seem like you really use the code much other than for getting the tree out.

Reply:
It's currently both a way of getting the tree in and also passing the code instance to the compiler and source generator consumers that also need state from the code. When we change them to no longer rely on RegexCode, this can trivially change along with it.